### PR TITLE
🎉 Destination Bigquery: added gcs upload option

### LINKED ISSUE #3549: Destination Bigquery returns 404 when uploading data + resumable
## Expected Behavior
Sync between Shopify and BQ works

## Current Behavior
It's not totally clear why this happens... 
BigQuery returns 404 when uploading data after running a few hours.

From Google Cloud API Page [here](https://cloud.google.com/bigquery/docs/reference/api-uploads#best-practices)
> Handle 404 Not Found and 410 Gone errors when doing resumable uploads by starting the entire upload over from the beginning.

## Logs
```
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - Exception in thread "main" com.google.cloud.bigquery.BigQueryException: 404 Not Found
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - PUT https://www.googleapis.com/upload/bigquery/v2/projects/earthbones-shopify/jobs?uploadType=resumable&upload_id=ABg5-UyhG8f29jrwEfJJ2mu4JTAIjUF_B3F5JvymoctXdtBQjQ29LsavMFtadgBiXThGjQcTtZAffIZJTf0940HdexnlUOtGew
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:115)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.write(HttpBigQueryRpc.java:705)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel$1.call(TableDataWriteChannel.java:59)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel$1.call(TableDataWriteChannel.java:54)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel.flushBuffer(TableDataWriteChannel.java:53)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDestination$RecordConsumer.acceptTracked(BigQueryDestination.java:358)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.accept(FailureTrackingAirbyteMessageConsumer.java:66)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:132)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:113)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDestination.main(BigQueryDestination.java:421)
```

## Steps to Reproduce
It's not clear how to reproduce. Maybe this happens when trying to sync large Shopify datasets to BQ.

## Severity of the bug for you
Very low / Low / **Medium** / High / Critical

## Airbyte Version
0.23.0-alpha

## Connector Version (if applicable)
0.2.2 Shopify

## Additional context
Slack [convo](https://airbytehq.slack.com/archives/C01MFR03D5W/p1621510734082900)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200368172141330) by [Unito](https://www.unito.io)

### LINKED ISSUE #5296: Destination Bigquery: rewrite connector to use Bulk upload instead of current one
## Current Behavior
Currently, we create all connectors and init all clients at the very beginning of connector creation. But it seems that google sdk big query connector may fail (with 404 response code) if we create it and wait for a long time (ex. 12 hours) while the source connector is reading data to migrate.   
Bulk loading (https://cloud.google.com/bigquery/docs/batch-loading-data) would allow us to stage the data on GCS entirely before loading it into bigquery. This is already implemented in Snowflake so we'd need something similar implemented for bigquery.

## Expected Behavior
[HelpFiles.zip](https://github.com/airbytehq/airbyte/files/6959635/HelpFiles.zip)
The connector should use Bulk upload type

## Logs
See attached archive

## ToDo as part of this ticket
1. **Update bigquery connector to use a Bulk loading** (collect data to GCS and then onClose move it to bigQuery using Bulk loadong)
Good to be also done:
2. Use the "destination bigquery creds" from lastpass to get a secret for testing
3. Run the performance test from attached archive to make sure that application doesn't return 404 anymore. (Create a new destination connector -> Wait for 12 + hours -> try write some message and stop container -> check that message appears on the cloud). For more details you may also check last comments from https://github.com/airbytehq/airbyte/issues/3549

Notes:
Some scoping required:
We already have some destination-gcs - how this can be re-used?
 Probably snowflake destination's implementation may be also useful to check before start working on this one

### PR DESCRIPTION
**Pull request by etsybaev**
## What
Currently, you may see some failures for big datasets and slow sources, i.e. if reading from source takes more than 10-12 hours. 
This is caused by the Google BigQuery SDK client limitations. For more details please check https://github.com/airbytehq/airbyte/issues/3549

## How
There are 2 available options now to upload data to bigquery `Standard` and `GCS Staging`.
- `Standard` is option to upload data directly from your source to BigQuery storage. This way is faster and requires less resources than GCS one.
- `GCS Uploading (CSV format)`. This is a newly introduced approach implemented in order to avoid the issue for big datasets mentioned above.
At the first step all data is uploaded to GCS bucket and then all moved to BigQuery at one shot stream by stream.
The destination-gcs connector is partially used under the hood here, so you may check its documentation for more details.

image
image

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by flagbug** [comment]
@etsybaev While looking for a workaround with the BigQuery connector failing with large datasets, I tried first manually uploading our tables to GCS but soon hit a limitation with the current GCS implementation as described here: https://github.com/airbytehq/airbyte/issues/5720

Since I'm assuming that you're reusing the same logic for the GCS writer I just wanted to let you know that this might also impact this PR.

**Review by irynakruk — COMMENTED** [review]
Nice pr!

**Review by irynakruk — APPROVED** [review]

File: airbyte-integrations/connectors/destination-bigquery/README.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+## Uploading options
+There are 2 available options to upload data to bigquery `Standard` and `GCS Staging`.
+- `Standard` is option to upload data directly from your source to BigQuery storage. This way is faster and requires less resources than GCS one.
```
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-bigquery/README.md:3]: I think we need to explain further about when to choose which option. Its not clear to me when should I choose `Standard` vs `GCS Uploading (CSV format)`

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 2]
```diff
@@ -239,12 +240,34 @@ public AirbyteMessageConsumer getConsumer(JsonNode config,
...
-    return getRecordConsumer(bigquery, writeConfigs, catalog, outputRecordCollector);
+    return getRecordConsumer(bigquery, writeConfigs, catalog, outputRecordCollector, isGcsUploadingMode, isKeepFilesInGcs);
+  }
+
+  private GcsCsvWriter initGcsWriter(GcsDestinationConfig gcsDestinationConfig,
+                                     ConfiguredAirbyteStream configuredStream)
+      throws IOException {
+    Timestamp uploadTimestamp = new Timestamp(System.currentTimeMillis());
+
+    AmazonS3 s3Client = GcsS3Helper.getGcsS3Client(gcsDestinationConfig);
```
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:269]: Why are we creating AmazonS3 client if we are using Google cloud storage for loading data

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 3]
```diff
@@ -239,12 +240,34 @@ public AirbyteMessageConsumer getConsumer(JsonNode config,
       }
       final WriteDisposition syncMode = getWriteDisposition(configStream.getDestinationSyncMode());
 
-      writeConfigs.put(AirbyteStreamNameNamespacePair.fromAirbyteSteam(stream),
-          new BigQueryWriteConfig(TableId.of(schemaName, tableName), TableId.of(schemaName, tmpTableName), writer, syncMode, schema));
+      if (isGcsUploadingMode) {
+        GcsDestinationConfig gcsDestinationConfig = GcsDestinationConfig
+            .getGcsDestinationConfig(BigQueryUtils.getGcsJsonNodeConfig(config));
+        GcsCsvWriter gcsCsvWriter = initGcsWriter(gcsDestinationConfig, configStream);
+        gcsCsvWriter.initialize();
```
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:247]: In the initialize method we are using AmazonS3 client and I am not sure I follow why is that

**Review by subodh1810 — CHANGES_REQUESTED** [review]
There is a major problem that I dont understand. Why are we using Amazon S3 client everywhere if we are trying to load data via GCS?

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java (hunk 2 above)
Comment (tuliren) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:269]: GCS is compatible with the Amazon S3 client. By reusing the S3 client, we can reuse the related code.

Comment (etsybaev) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:269]: Hi @subodh1810.  I just partially re-used our already existing destination-gcs module that had been created before I started working on this ticket.  I had some conversation with @tuliren  and he confirmed that S3 client was also used for destination-gcs as it mostly works with both except for some minor cases.

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java (hunk 3 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:247]: Hi @subodh1810. The answer here is similar to the previous comment. I partially re-used the already existing destination-gcs module (GCS destination CSV writer in particular). But destination-gcs had been implemented re-using already existed modules from destination-S3 as far as I understood. The reason is that the Amazon S3 client in most of the cases works for both GCS and S3 storages. It's better to ask @tuliren for details. 
So the idea is the next:
1. Re-using the existing destination-gcs (CSV writer) we upload data to GCS bucket
2. Using the native Bigquery writer we create a native job to migrate CSV formatted data from bucket to BigQuery

File: airbyte-integrations/connectors/destination-bigquery/README.md (hunk 1 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-bigquery/README.md:3]: Updated, thanks

**Review by subodh1810 — APPROVED** [review]
Can you just add a java doc explaining that Amazon s3 client works with GCS as well

**Comment by sherifnada** [comment]
@etsybaev see #5959 - why did we change the emitted_at column type? is that a change we can revert?

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 4]
```diff
@@ -73,17 +77,13 @@
...
-  static final String CONFIG_CREDS = "credentials_json";
-  static final String BIG_QUERY_CLIENT_CHUNK_SIZE = "big_query_client_buffer_size_mb";
 
   private static final com.google.cloud.bigquery.Schema SCHEMA = com.google.cloud.bigquery.Schema.of(
       Field.of(JavaBaseConstants.COLUMN_NAME_AB_ID, StandardSQLTypeName.STRING),
-      Field.of(JavaBaseConstants.COLUMN_NAME_DATA, StandardSQLTypeName.STRING),
-      Field.of(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, StandardSQLTypeName.TIMESTAMP));
+      // GCS works with only date\datetime formats, so need to have it a string for a while
+      // https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-csv#data_types
+      Field.of(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, StandardSQLTypeName.STRING),
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:85]: The type of the column when creating the schema of the table was changed from timestamp to STRING

But when formatting the records to insert, they are still typed as timestamp?
https://github.com/airbytehq/airbyte/blob/888e9abee4b653054ae7fc80b860e966120f79b3/airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryRecordConsumer.java#L153

Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:85]: ok so there was an issue for this here https://github.com/airbytehq/airbyte/issues/5959

**Pull request merged**